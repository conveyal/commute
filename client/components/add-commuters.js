/* globals FileReader */

import {csvParse} from 'd3-dsv'
import React, {Component, PropTypes} from 'react'
import {Accordion, Button, Col, Grid, Panel, Row} from 'react-bootstrap'
import {BootstrapTable, TableHeaderColumn} from 'react-bootstrap-table'
import Dropzone from 'react-dropzone'
import {Link} from 'react-router'
import uuid from 'uuid'

import FieldGroup from './fieldgroup'
import Icon from './icon'

export default class AddCommuters extends Component {
  static propTypes = {
    // dispatch
    create: PropTypes.func,
    append: PropTypes.func,

    // props
    appendMode: PropTypes.bool,
    group: PropTypes.object,
    organizationId: PropTypes.string.isRequired
  }

  componentWillMount () {
    if (this.props.appendMode) {
      this.setState({
        name: this.props.group.name,
        existingCommuters: this.props.group.commuters
      })
    } else {
      this.state = {}
    }
  }

  handleChange = (name, event) => {
    this.setState({ [name]: event.target.value })
  }

  handleSubmit = () => {
    const {append, appendMode, create, group, organizationId} = this.props
    if (appendMode) {
      append(this.state.newCommuters, group.id, organizationId)
    } else {
      const newGroup = {...this.state}
      newGroup.commuters = newGroup.newCommuters
      delete newGroup.newCommuters
      create(newGroup, organizationId)
    }
  }

  makeCommuterTable (commuters) {
    return (
      <BootstrapTable data={commuters}>
        <TableHeaderColumn dataField='id' isKey hidden />
        <TableHeaderColumn dataField='name'>Name</TableHeaderColumn>
        <TableHeaderColumn dataField='email'>Email</TableHeaderColumn>
        <TableHeaderColumn dataField='address'>Address</TableHeaderColumn>
      </BootstrapTable>
    )
  }

  onDrop = (files) => {
    const r = new FileReader()

    r.onload = (e) => {
      const newCommuters = csvParse(e.target.result, (row) => {
        const {address, email, name} = row
        const id = row.id || uuid.v4()
        // TODO: parse more field possibilities (first name, last name, etc)
        return {address, email, id, name}
      })

      this.setState({newCommuters})
    }

    // TODO: handle multiple files
    r.readAsText(files[0])
  }

  render () {
    const {appendMode, organizationId} = this.props
    const groupName = this.state.name
    const showAccordion = !!(this.state.existingCommuters || this.state.newCommuters)
    return (
      <Grid>
        <Row>
          <Col xs={12}>
            <h3>
              <span>{appendMode ? 'Add Commuters to Group' : 'Create Commuter Group'}</span>
              <Button className='pull-right'>
                <Link to={`/organizations/${organizationId}`}><Icon type='arrow-left' />Back</Link>
              </Button>
            </h3>
            <form>
              <FieldGroup
                label='Name'
                name='name'
                onChange={this.handleChange}
                placeholder='Enter name'
                type='text'
                value={groupName}
                />
              <Dropzone
                accept='text/csv'
                className='dropzone'
                multiple={false}
                onDrop={this.onDrop}
                >
                <div>Try dropping a csv file here, or click to select files to upload.  Make sure the csv file contains the headers: name, email and address.</div>
              </Dropzone>
              {showAccordion &&
                <Accordion>
                  {!!(this.state.existingCommuters) &&
                    <Panel
                      header={`${this.state.existingCommuters.length} Existing Commuters`}
                      bsStyle='info'
                      eventKey='1'
                      >
                      {this.makeCommuterTable(this.state.existingCommuters)}
                    </Panel>
                  }
                  {!!(this.state.newCommuters) &&
                    <Panel
                      header={`${this.state.newCommuters.length} New Commuters`}
                      bsStyle='success'
                      eventKey='2'
                      >
                      {this.makeCommuterTable(this.state.newCommuters)}
                    </Panel>
                  }
                </Accordion>
              }
              <Button
                bsStyle='success'
                onClick={this.handleSubmit}
                >
                {appendMode ? 'Append' : 'Create'}
              </Button>
            </form>
          </Col>
        </Row>
      </Grid>
    )
  }
}
