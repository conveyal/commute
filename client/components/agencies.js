import React, {Component, PropTypes} from 'react'
import {Button, ButtonGroup, Col, Grid, Row} from 'react-bootstrap'
import {BootstrapTable, TableHeaderColumn} from 'react-bootstrap-table'
import {Link} from 'react-router'

import ButtonLink from './button-link'
import Icon from './icon'
import {messages} from '../utils/env'
import {arrayCountRenderer} from '../utils/table'
import {actUponConfirmation} from '../utils/ui'

export default class Agencies extends Component {
  static propTypes = {
    // dispatch
    deleteAgency: PropTypes.func.isRequired,
    loadAgencies: PropTypes.func.isRequired,

    // props
    agencies: PropTypes.array.isRequired
  }

  componentWillMount () {
    this.props.loadAgencies()
  }

  _onDeleteAgencyClick (agency) {
    const doDelete = () => { this.props.deleteAgency(agency) }
    actUponConfirmation(messages.agency.deleteConfirmation, doDelete)
  }

  _toolsRenderer = (cell, row) => {
    return <ButtonGroup>
      <ButtonLink
        bsStyle='warning'
        to={`/agency/${row._id}/edit`}
        >
        Edit
      </ButtonLink>
      <Button bsStyle='danger' onClick={this._onDeleteAgencyClick.bind(this, row)}>Delete</Button>
    </ButtonGroup>
  }

  render () {
    const {agencies} = this.props
    return (
      <Grid>
        <Row>
          <Col xs={12}>
            <h2>Agencies
              <ButtonLink
                className='pull-right'
                to='/agency/create'
                >
                <span>Create a new agency</span>
                <Icon type='flag' />
              </ButtonLink>
            </h2>
            <p>An agency is a collection of organizations <Icon type='shield' />.</p>
            <BootstrapTable data={agencies}>
              <TableHeaderColumn dataField='_id' isKey hidden />
              <TableHeaderColumn dataField='name' dataFormat={nameRenderer}>Name</TableHeaderColumn>
              <TableHeaderColumn dataField='organizations' dataFormat={arrayCountRenderer}>Organizations</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._toolsRenderer}>Tools</TableHeaderColumn>
            </BootstrapTable>
          </Col>
        </Row>
      </Grid>
    )
  }
}

function nameRenderer (cell, row) {
  return <Link to={`/agency/${row._id}`}>{cell}</Link>
}
