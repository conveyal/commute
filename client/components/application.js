import React, {Component, PropTypes} from 'react'

import Footer from '../components/footer'
import Navigation from '../containers/navigation'

export default class Application extends Component {
  static propTypes = {
    // actions
    refreshUserToken: PropTypes.func.isRequired
  }

  componentWillMount () {
    this.props.refreshUserToken()
  }

  render () {
    const {children} = this.props
    return (
      <div>
        <Navigation />
        {children}
        <Footer />
      </div>
    )
  }
}
