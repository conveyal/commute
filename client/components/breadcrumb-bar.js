import React, {Component, PropTypes} from 'react'
import {Col, Grid, Row} from 'react-bootstrap'
import {Link} from 'react-router'

export default class BreadcrumbBar extends Component {
  static propTypes = {
    // props
    multiSite: PropTypes.object.isRequired,
    site: PropTypes.object.isRequired
  }

  _makeNavItems () {
    const navItems = [{
      href: '/',
      name: 'Home'
    }]

    const path = process.env.NODE_ENV === 'test' ? window.fakePath : window.location.pathname

    const appendEntity = (id, name, urlBase, isActive) => {
      if (isActive) {
        navItems.push({
          active: true,
          name
        })
      } else {
        navItems.push({
          href: urlBase + id,
          name
        })
      }
    }

    const appendMultiSite = (multiSiteId, isActive) => {
      const multiSite = this.props.multiSite[multiSiteId]
      appendEntity(multiSiteId, multiSite ? multiSite.name : '?', '/multi-site/', isActive)
    }

    const appendSite = (siteId, isActive) => {
      const site = this.props.site[siteId]
      appendEntity(siteId, site ? site.name : '?', '/site/', isActive)
    }

    if (path === '/' || path === '/login') {
      // do nothing for home or login
    } else if (path.match(/\/multi-site\/create$/)) {
      // Create Site View
      navItems.push({
        active: true,
        name: 'Create Multi-Site Analysis'
      })
    } else if (path.match(/\/multi-site\/[\w-]+$/)) {
      // Site View
      appendMultiSite(path.match(/\/multi-site\/([\w-]+)/)[1], true)
    } else if (path.match(/\/site\/create$/)) {
      // Create Site View
      navItems.push({
        active: true,
        name: 'Create Site'
      })
    } else if (path.match(/\/site\/[\w-]+$/)) {
      // Site View
      appendSite(path.match(/\/site\/([\w-]+)/)[1], true)
    } else if (path.match(/\/site\/[\w-]+\/bulk-add-commuters$/)) {
      // Create Commuter View
      appendSite(path.match(/\/site\/([\w-]+)/)[1])
      navItems.push({
        active: true,
        name: 'Bulk Add Commuters'
      })
    } else if (path.match(/\/site\/[\w-]+\/commuter\/create$/)) {
      // Create Commuter View
      appendSite(path.match(/\/site\/([\w-]+)/)[1])
      navItems.push({
        active: true,
        name: 'Create New Commuter'
      })
    } else if (path.match(/\/site\/[\w-]+\/commuter\/[\w-]+\/edit$/)) {
      // Create Commuter View
      appendSite(path.match(/\/site\/([\w-]+)/)[1])
      navItems.push({
        active: true,
        name: 'Edit Commuter'
      })
    } else if (path.match(/\/site\/[\w-]+\/edit$/)) {
      // Edit Site View
      appendSite(path.match(/\/site\/([\w-]+)/)[1])
      navItems.push({
        active: true,
        name: 'Edit Site'
      })
    } else {
      console.error(`Path not matched for breadcrumb generation: ${path}`)
    }
    return navItems
  }

  render () {
    const navItems = this._makeNavItems()
    return (
      <Grid>
        <Row>
          <Col xs={12}>
            <ol role='navigation' aria-label='breadcrumbs' className='breadcrumb'>
              {navItems.map((item, idx) => {
                return (
                  <li
                    key={idx}
                    >
                    {item.active ? item.name : <Link to={item.href}>{item.name}</Link>}
                  </li>
                )
              })}
            </ol>
          </Col>
        </Row>
      </Grid>
    )
  }
}
