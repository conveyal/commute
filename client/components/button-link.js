import invariant from 'invariant'
import React, {Component, PropTypes} from 'react'
import {Button} from 'react-bootstrap'

export default class ButtonLink extends Component {
  static propTypes = {
    to: PropTypes.string.isRequired
  }

  static contextTypes = {
    router: PropTypes.object
  }

  _handleClick = (event) => {
    // copied from react-router
    if (this.props.onClick) this.props.onClick(event)

    if (event.defaultPrevented) return

    const { router } = this.context
    invariant(
      router,
      '<ButtonLink>s rendered outside of a router context cannot navigate.'
    )

    if (isModifiedEvent(event) || !isLeftClickEvent(event)) return

    // If target prop is set (e.g. to "_blank"), let browser handle link.
    /* istanbul ignore if: untestable with Karma */
    if (this.props.target) return

    event.preventDefault()

    router.push(resolveToLocation(this.props.to, router))
  }

  render () {
    const {children, ...buttonProps} = this.props
    return (
      <Button
        {...buttonProps}
        onClick={this._handleClick}
        >
        {children}
      </Button>
    )
  }
}

function isLeftClickEvent (event) {
  return event.button === 0
}

function isModifiedEvent (event) {
  return !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey)
}

function resolveToLocation (to, router) {
  return typeof to === 'function' ? to(router.location) : to
}
