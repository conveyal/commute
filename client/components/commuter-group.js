import {Browser, latLngBounds} from 'leaflet'
import React, {Component, PropTypes} from 'react'
import {Button, Col, Grid, Row} from 'react-bootstrap'
import {BootstrapTable, TableHeaderColumn} from 'react-bootstrap-table'
import {Map as LeafletMap, TileLayer} from 'react-leaflet'
import ClusterLayer from 'react-leaflet-cluster-layer'
import {Link} from 'react-router'

import Icon from './icon'
import {messages, settings} from '../utils/env'
import {actUponConfirmation} from '../utils/ui'

export default class CommuterGroup extends Component {
  static propTypes = {
    // dispatch
    deleteCommuter: PropTypes.func,
    deleteGroup: PropTypes.func,
    update: PropTypes.func,

    // props
    organizationId: PropTypes.string.isRequired,
    group: PropTypes.object.isRequired
  }

  _commuterToolsRenderer = (cell, row) => {
    const groupId = this.props.group.id
    const organizationId = this.props.organizationId
    const doDelete = () => this.props.deleteCommuter({ commuterId: row.id, groupId, organizationId })
    const onClick = () => actUponConfirmation(messages.commuter.deleteConfirmation, doDelete)
    return <div>
      <Button bsStyle='warning'>
        <Link to={`/organizations/${organizationId}/groups/${groupId}/commuters/${row.id}/edit`}>Edit</Link>
      </Button>
      <Button bsStyle='danger' onClick={onClick}>Delete</Button>
    </div>
  }

  handleDelete = () => {
    const doDelete = () => this.props.deleteGroup(this.state.group.id, this.props.organizationId)
    actUponConfirmation(messages.organization.deleteConfirmation, doDelete)
  }

  render () {
    const {group, organizationId} = this.props
    const {allAddressesGeocoded, commuters} = group
    const groupName = group.name
    const {bounds, markers, position, zoom} = mapCommuters(allAddressesGeocoded, commuters)
    return (
      <Grid>
        <Row>
          <Col xs={12} className='group-header'>
            <h3>
              <span>{groupName}</span>
              <Button className='pull-right'>
                <Link to={`/organizations/${organizationId}`}><Icon type='arrow-left' />Back</Link>
              </Button>
            </h3>
          </Col>
          {allAddressesGeocoded &&
            <Col xs={12} style={{height: '400px'}}>
              <LeafletMap center={position} bounds={bounds} zoom={zoom}>
                <TileLayer
                  url={Browser.retina &&
                    process.env.LEAFLET_RETINA_URL
                    ? process.env.LEAFLET_RETINA_URL
                    : process.env.LEAFLET_TILE_URL}
                  attribution={process.env.LEAFLET_ATTRIBUTION}
                  />
                <ClusterLayer
                  markers={markers}
                  clusterComponent={ClusterComponent}
                  />
              </LeafletMap>
            </Col>
          }
          {!allAddressesGeocoded &&
            <Col xs={12} className='group-geocode-progress'>
              <p>Geocoding commuter addresses...</p>
            </Col>
          }
        </Row>
        <Row className='group-content'>
          <Col xs={12}>
            <BootstrapTable data={commuters}>
              <TableHeaderColumn dataField='id' isKey hidden />
              <TableHeaderColumn dataField='name'>Name</TableHeaderColumn>
              <TableHeaderColumn dataField='address'>Address</TableHeaderColumn>
              <TableHeaderColumn dataField='email'>Email</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._commuterToolsRenderer}>Tools</TableHeaderColumn>
            </BootstrapTable>
          </Col>
        </Row>
      </Grid>
    )
  }
}

class ClusterComponent extends React.Component {
  render () {
    const style = {
      border: 'solid 2px darkgrey',
      borderRadius: '8px',
      backgroundColor: 'white',
      padding: '1em',
      textAlign: 'center'
    }
    const cluster = this.props.cluster

    if (cluster.markers.length === 1) {
      return (
        <div style={style} >{cluster.markers[0].name}</div>
      )
    }

    return (
      <div style={style}>{cluster.markers.length} items</div>
    )
  }
}

const mapCommuters = (allAddressesGeocoded, commuters) => {
  if (commuters.length === 0 || !allAddressesGeocoded) {
    return {
      markers: [],
      position: settings.map.focus,
      zoom: 8
    }
  } else if (commuters.length === 1) {
    return {
      markers: [{
        id: commuters[0].id,
        name: commuters[0].name,
        position: {
          lat: commuters[0].lat,
          lng: commuters[0].lng
        }
      }],
      position: settings.map.focus,
      zoom: 8
    }
  }
  const markers = []
  const firstLL = [commuters[0].lat, commuters[0].lng]
  const bounds = latLngBounds([firstLL, firstLL])
  commuters.forEach((commuter) => {
    const {id, lat, lng, name} = commuter
    markers.push({
      id,
      name,
      position: { lat, lng }
    })
    bounds.extend([lat, lng])
  })
  return {bounds, markers}
}
