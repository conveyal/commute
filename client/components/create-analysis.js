import React, {Component, PropTypes} from 'react'
import {Button, Col, ControlLabel, FormGroup, Grid, Row} from 'react-bootstrap'
import {Link} from 'react-router'
import Select from 'react-select'

import Icon from './icon'

export default class CreateAnalysis extends Component {
  static propTypes = {
    // dispatch
    create: PropTypes.func,

    // props
    organization: PropTypes.object.isRequired
  }

  constructor (props) {
    super(props)
    this.state = {}
  }

  _handleGroupChange = (event) => {
    const groupId = event.value
    this.setState({
      commuters: this.props.organization.groupsById[groupId].commuters,
      groupId
    })
  }

  _handleSiteChange = (event) => {
    this.setState({ siteId: event.value })
  }

  _handleSubmit = () => {
    this.props.create(this.state, this.props.organization.id)
  }

  render () {
    const {groups, id: organizationId, sites} = this.props.organization
    return (
      <Grid>
        <Row>
          <Col xs={12}>
            <h3>
              <span>Create Analysis</span>
              <Button className='pull-right'>
                <Link to={`/organizations/${organizationId}`}><Icon type='arrow-left' />Back</Link>
              </Button>
            </h3>
            <form>
              <FormGroup controlId='site-control'>
                <ControlLabel>Site</ControlLabel>
                <Select
                  onChange={this._handleSiteChange}
                  options={sites.map((site) => { return {value: site.id, label: site.name} })}
                  placeholder='Select a Site...'
                  value={this.state.siteId}
                  />
              </FormGroup>
              <FormGroup controlId='group-control'>
                <ControlLabel>Group</ControlLabel>
                <Select
                  onChange={this._handleGroupChange}
                  options={groups.map((group) => { return {value: group.id, label: group.name} })}
                  placeholder='Select a Commuter Group...'
                  value={this.state.groupId}
                  />
              </FormGroup>
              <Button
                bsStyle='success'
                onClick={this._handleSubmit}
                >
                Create
              </Button>
            </form>
          </Col>
        </Row>
      </Grid>
    )
  }
}
