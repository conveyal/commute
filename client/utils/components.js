import {settings} from './env'

/**
 * Calculate number of elements in array less than or equal to target values
 *
 * @param  {Number[]} arr     A sorted array of numbers
 * @param  {[Number]} target  The number to be less than or equal to
 * @return {int}              The number of elements in the array less than or equal to the target value
 */
export function calcNumLessThan (arr, target) {
  // do binary search
  let left = 0
  let right = arr.length - 1
  let mid

  while (left <= right) {
    mid = Math.floor((left + right) / 2)
    if (arr[mid] <= target) {
      left = mid + 1
    } else {
      right = mid - 1
    }
  }

  return left
}

/**
 * Helper function to get initial series data for drawing graphs
 *
 * @return {Obj[]} Array of objects with series template
 */
export function getInitialSeries () {
  const modes = Object.keys(settings.modeDisplay)
  const series = []

  modes.forEach((mode) => {
    series.push(Object.assign({disabled: false, mode}, settings.modeDisplay[mode]))
  })

  return series
}

/**
 * Humanize distances
 * // TODO: more humanization
 *
 * @param  {float} d Distance to humanize
 * @return {string} humanized distance
 */
export function humanizeDistance (d) {
  return `${d} miles`
}
