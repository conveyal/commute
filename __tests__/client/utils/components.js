/* globals describe, expect, it */

import * as components from '../../../client/utils/components'

describe('utils > components', () => {
  it('humanizeDistance should work', () => {
    expect(components.humanizeDistance(123)).toMatchSnapshot()
  })

  describe('> calcNumLessThan', () => {
    const testCases = [{
      arr: [],
      target: 3,
      expected: 0
    }, {
      arr: [1],
      target: 3,
      expected: 1
    }, {
      arr: [4],
      target: 3,
      expected: 0
    }, {
      arr: [1, 2, 3, 4],
      target: 0,
      expected: 0
    }, {
      arr: [1, 2, 3, 4],
      target: 5,
      expected: 4
    }, {
      arr: [1, 2, 3, 4],
      target: 2,
      expected: 2
    }, {
      arr: [1, 2, 3, 4],
      target: 3,
      expected: 3
    }, {
      arr: [1, 2, 3],
      target: 2,
      expected: 2
    }, {
      arr: [1, 1, 1, 1],
      target: 1,
      expected: 4
    }, {
      arr: [1, 1, 1, 1],
      target: 0,
      expected: 0
    }, {
      arr: [1, 2, 2, 2, 5],
      target: 2,
      expected: 4
    }]

    testCases.forEach((tc, idx) => {
      it(`should compute correct amount for test case ${idx}`, () => {
        expect(components.calcNumLessThan(tc.arr, tc.target)).toBe(tc.expected)
      })
    })
  })
})
