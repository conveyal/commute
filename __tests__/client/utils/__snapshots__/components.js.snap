exports[`utils > components humanizeDistance should work 1`] = `"123 miles"`;
